File: src/management/LibraryManager.cpp
// management/LibraryManager.cpp
#include "LibraryManager.h"
#include "../utils/FileHandler.h"
#include <iostream>
#include <algorithm>

// Initialize static instance
LibraryManager* LibraryManager::instance = nullptr;

// ============ SINGLETON ============

LibraryManager::LibraryManager() {
    bookTree = new BookBST();
    userMap = new UserHashMap();
    transactionList = new TransactionList();
    searchEngine = new SearchEngine();
    authManager = nullptr;
    
    // Link search engine to book tree
    searchEngine->setBookTree(bookTree);
}

LibraryManager* LibraryManager::getInstance() {
    if (instance == nullptr) {
        instance = new LibraryManager();
    }
    return instance;
}

LibraryManager::~LibraryManager() {
    delete bookTree;
    delete userMap;
    delete transactionList;
    delete searchEngine;
    // Don't delete authManager (owned externally)
}

void LibraryManager::setAuthManager(AuthManager* auth) {
    authManager = auth;
}

// ============ ADMIN OPERATIONS - BOOK MANAGEMENT ============

bool LibraryManager::addBook(const string& isbn, const string& title, 
                             const string& author, int quantity) {
    // Check admin privileges
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    return addBookInternal(isbn, title, author, quantity);
}

// FIX #4: Private helper without auth check (for initialization)
bool LibraryManager::addBookInternal(const string& isbn, const string& title, 
                                    const string& author, int quantity) {
    // Validate ISBN
    if (!isISBNValid(isbn)) {
        cout << "Error: Invalid ISBN format." << endl;
        return false;
    }
    
    // Check if book already exists
    Book* existing = bookTree->search(isbn);
    if (existing != nullptr) {
        cout << "Error: Book with ISBN " << isbn << " already exists." << endl;
        return false;
    }
    
    // Create and insert book
    Book newBook(isbn, title, author, quantity);
    bookTree->insert(newBook);
    
    // Update search indices
    searchEngine->addBookToIndex(newBook);
    
    cout << "Success: Book added successfully." << endl;
    return true;
}

bool LibraryManager::removeBook(const string& isbn) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    Book* book = bookTree->search(isbn);
    if (book == nullptr) {
        cout << "Error: Book not found." << endl;
        return false;
    }
    
    // Check if any copies are currently borrowed
    int borrowed = book->getQuantity() - book->getAvailableCopies();
    if (borrowed > 0) {
        cout << "Error: Cannot remove book. " << borrowed 
             << " copies are currently borrowed." << endl;
        return false;
    }
    
    // Remove from indices first
    searchEngine->removeBookFromIndex(isbn);
    
    // Remove from tree
    if (bookTree->remove(isbn)) {
        cout << "Success: Book removed successfully." << endl;
        return true;
    }
    
    return false;
}

bool LibraryManager::updateBookDetails(const string& isbn, const string& newTitle, 
                                       const string& newAuthor) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    Book* book = bookTree->search(isbn);
    if (book == nullptr) {
        cout << "Error: Book not found." << endl;
        return false;
    }
    
    // Update book details
    Book updatedBook(isbn, newTitle, newAuthor, book->getQuantity());
    updatedBook.setAvailableCopies(book->getAvailableCopies());
    
    // Remove old indices
    searchEngine->removeBookFromIndex(isbn);
    
    // Update in tree
    bookTree->insert(updatedBook);  // Will replace existing
    
    // Add new indices
    searchEngine->addBookToIndex(updatedBook);
    
    cout << "Success: Book details updated." << endl;
    return true;
}

bool LibraryManager::updateBookQuantity(const string& isbn, int newQuantity) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    Book* book = bookTree->search(isbn);
    if (book == nullptr) {
        cout << "Error: Book not found." << endl;
        return false;
    }
    
    int borrowed = book->getQuantity() - book->getAvailableCopies();
    
    if (newQuantity < borrowed) {
        cout << "Error: Cannot reduce quantity below borrowed copies (" 
             << borrowed << ")." << endl;
        return false;
    }
    
    int difference = newQuantity - book->getQuantity();
    book->setQuantity(newQuantity);
    book->setAvailableCopies(book->getAvailableCopies() + difference);
    
    cout << "Success: Book quantity updated." << endl;
    return true;
}

vector<Book*> LibraryManager::getAllBooks() {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<Book*>();
    }
    return bookTree->getAllBooksSorted();
}

vector<Book*> LibraryManager::getAvailableBooks() {
    return searchEngine->searchAvailableBooks();
}

// ============ ADMIN OPERATIONS - USER MANAGEMENT ============

vector<User*> LibraryManager::getAllUsers() {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<User*>();
    }
    return userMap->getAllUsers();
}

bool LibraryManager::removeUser(const string& userID) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    User* user = userMap->searchByID(userID);
    if (user == nullptr) {
        cout << "Error: User not found." << endl;
        return false;
    }
    
    // Check if user has borrowed books
    if (user->getBorrowedCount() > 0) {
        cout << "Error: Cannot remove user. User has " 
             << user->getBorrowedCount() << " borrowed books." << endl;
        return false;
    }
    
    if (userMap->remove(userID)) {
        cout << "Success: User removed successfully." << endl;
        return true;
    }
    
    return false;
}

bool LibraryManager::deactivateUser(const string& userID) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    User* user = userMap->searchByID(userID);
    if (user == nullptr) {
        cout << "Error: User not found." << endl;
        return false;
    }
    
    user->setActive(false);
    cout << "Success: User account deactivated." << endl;
    return true;
}

bool LibraryManager::activateUser(const string& userID) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        cout << "Access Denied: Admin privileges required." << endl;
        return false;
    }
    
    User* user = userMap->searchByID(userID);
    if (user == nullptr) {
        cout << "Error: User not found." << endl;
        return false;
    }
    
    user->setActive(true);
    cout << "Success: User account activated." << endl;
    return true;
}

User* LibraryManager::getUserDetails(const string& userID) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return nullptr;
    }
    return userMap->searchByID(userID);
}

// ============ ADMIN OPERATIONS - TRANSACTIONS ============

vector<Transaction*> LibraryManager::getAllTransactions() {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<Transaction*>();
    }
    return transactionList->getAll();
}

vector<Transaction*> LibraryManager::getUserTransactions(const string& userID) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<Transaction*>();
    }
    return transactionList->getByUserID(userID);
}

vector<Transaction*> LibraryManager::getBookTransactions(const string& isbn) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<Transaction*>();
    }
    return transactionList->getByISBN(isbn);
}

vector<Transaction*> LibraryManager::getRecentTransactions(int count) {
    if (authManager == nullptr || !authManager->isAdmin()) {
        return vector<Transaction*>();
    }
    return transactionList->getRecent(count);
}

// ============ ADMIN OPERATIONS - STATISTICS ============

int LibraryManager::getTotalBooks() {
    return bookTree->getCount();
}

int LibraryManager::getTotalAvailableBooks() {
    vector<Book*> books = bookTree->getAllBooksSorted();
    int availableCount = 0;
    for (Book* book : books) {
        availableCount += book->getAvailableCopies();
    }
    return availableCount;
}

int LibraryManager::getTotalUsers() {
    return userMap->getCount();
}

int LibraryManager::getTotalTransactions() {
    return transactionList->getCount();
}

int LibraryManager::getActiveUsersCount() {
    vector<User*> users = userMap->getAllUsers();
    int activeCount = 0;
    for (User* user : users) {
        if (user->isActive()) {
            activeCount++;
        }
    }
    return activeCount;
}

// ============ USER OPERATIONS - SEARCH ============

vector<Book*> LibraryManager::searchBooksByTitle(const string& title) {
    return searchEngine->searchByTitle(title);
}

vector<Book*> LibraryManager::searchBooksByAuthor(const string& author) {
    return searchEngine->searchByAuthor(author);
}

vector<Book*> LibraryManager::searchBooksByKeyword(const string& keyword) {
    return searchEngine->searchByKeyword(keyword);
}

Book* LibraryManager::searchBookByISBN(const string& isbn) {
    return searchEngine->searchByISBN(isbn);
}

// ============ USER OPERATIONS - BORROW & RETURN ============

bool LibraryManager::borrowBook(const string& isbn) {
    // Check if user is logged in
    if (authManager == nullptr || !authManager->isUser()) {
        cout << "Access Denied: Please login as user." << endl;
        return false;
    }
    
    User* currentUser = authManager->getCurrentUser();
    if (currentUser == nullptr) {
        cout << "Error: User session invalid." << endl;
        return false;
    }
    
    // Check if user can borrow
    if (!currentUser->canBorrow()) {
        cout << "Error: Borrowing limit reached (" << MAX_BORROW_LIMIT 
             << " books)." << endl;
        return false;
    }
    
    // Find book
    Book* book = bookTree->search(isbn);
    if (book == nullptr) {
        cout << "Error: Book not found." << endl;
        return false;
    }
    
    // Check if book is available
    if (!book->isAvailable()) {
        cout << "Error: Book is not available." << endl;
        return false;
    }
    
    // Check if user already has this book
    if (currentUser->hasBorrowedBook(isbn)) {
        cout << "Error: You have already borrowed this book." << endl;
        return false;
    }
    
    // Perform borrowing
    if (book->borrowBook()) {
        currentUser->addBorrowedBook(isbn);
        
        // Create transaction record
        Transaction* trans = new Transaction(
            currentUser->getUserID(),
            isbn,
            "BORROW",
            currentUser->getFullName(),
            book->getTitle()
        );
        transactionList->append(trans);
        
        cout << "Success: Book borrowed successfully." << endl;
        cout << "Books borrowed: " << currentUser->getBorrowedCount() 
             << "/" << MAX_BORROW_LIMIT << endl;
        return true;
    }
    
    return false;
}

bool LibraryManager::returnBook(const string& isbn) {
    // Check if user is logged in
    if (authManager == nullptr || !authManager->isUser()) {
        cout << "Access Denied: Please login as user." << endl;
        return false;
    }
    
    User* currentUser = authManager->getCurrentUser();
    if (currentUser == nullptr) {
        cout << "Error: User session invalid." << endl;
        return false;
    }
    
    // Check if user has borrowed this book
    if (!currentUser->hasBorrowedBook(isbn)) {
        cout << "Error: You have not borrowed this book." << endl;
        return false;
    }
    
    // Find book
    Book* book = bookTree->search(isbn);
    if (book == nullptr) {
        cout << "Error: Book not found." << endl;
        return false;
    }
    
    // Perform return
    if (book->returnBook()) {
        currentUser->removeBorrowedBook(isbn);
        
        // Create transaction record
        Transaction* trans = new Transaction(
            currentUser->getUserID(),
            isbn,
            "RETURN",
            currentUser->getFullName(),
            book->getTitle()
        );
        transactionList->append(trans);
        
        cout << "Success: Book returned successfully." << endl;
        cout << "Books borrowed: " << currentUser->getBorrowedCount() 
             << "/" << MAX_BORROW_LIMIT << endl;
        return true;
    }
    
    return false;
}

vector<Book*> LibraryManager::getMyBorrowedBooks() {
    if (authManager == nullptr || !authManager->isUser()) {
        return vector<Book*>();
    }
    
    User* currentUser = authManager->getCurrentUser();
    if (currentUser == nullptr) {
        return vector<Book*>();
    }
    
    vector<Book*> borrowedBooks;
    set<string> borrowedISBNs = currentUser->getBorrowedISBNs();
    
    for (const string& isbn : borrowedISBNs) {
        Book* book = bookTree->search(isbn);
        if (book != nullptr) {
            borrowedBooks.push_back(book);
        }
    }
    
    return borrowedBooks;
}

vector<Transaction*> LibraryManager::getMyTransactions() {
    if (authManager == nullptr || !authManager->isUser()) {
        return vector<Transaction*>();
    }
    
    User* currentUser = authManager->getCurrentUser();
    if (currentUser == nullptr) {
        return vector<Transaction*>();
    }
    
    return transactionList->getByUserID(currentUser->getUserID());
}

// ============ DATA PERSISTENCE ============

bool LibraryManager::saveAllData() {
    bool success = true;
    
    success &= FileHandler::saveBooks(BOOKS_FILE, bookTree);
    success &= FileHandler::saveUsers(USERS_FILE, userMap);
    success &= FileHandler::saveTransactions(TRANSACTIONS_FILE, transactionList);
    
    if (success) {
        cout << "Success: All data saved successfully." << endl;
    } else {
        cout << "Warning: Some data may not have been saved." << endl;
    }
    
    return success;
}

bool LibraryManager::loadAllData() {
    bool success = true;
    
    success &= FileHandler::loadBooks(BOOKS_FILE, bookTree);
    success &= FileHandler::loadUsers(USERS_FILE, userMap);
    success &= FileHandler::loadTransactions(TRANSACTIONS_FILE, transactionList);
    
    // Rebuild search indices after loading books
    searchEngine->buildIndices();
    
    if (success) {
        cout << "Success: All data loaded successfully." << endl;
    }
    
    return success;
}

void LibraryManager::initializeSampleData() {
    cout << "Initializing sample data..." << endl;
    
    // Add sample books
    addBookInternal("978-0-13-468599-1", "The C++ Programming Language", 
                    "Bjarne Stroustrup", 5);
    addBookInternal("978-0-321-56384-2", "Effective C++", 
                    "Scott Meyers", 3);
    addBookInternal("978-0-262-03384-8", "Introduction to Algorithms", 
                    "Thomas Cormen", 4);
    addBookInternal("978-0-201-35088-5", "Data Structures and Algorithms", 
                    "Alfred Aho", 3);
    addBookInternal("978-0-672-32692-7", "Data Structures Using C++", 
                    "D.S. Malik", 4);
    addBookInternal("978-0-201-63361-0", "Design Patterns", 
                    "Gang of Four", 2);
    addBookInternal("978-0-596-00927-5", "Head First Design Patterns", 
                    "Eric Freeman", 3);
    addBookInternal("978-0-201-89683-1", "The Art of Computer Programming Vol 1", 
                    "Donald Knuth", 2);
    addBookInternal("978-0-13-110362-7", "The C Programming Language", 
                    "Brian Kernighan", 5);
    addBookInternal("978-0-134-68599-4", "Clean Code", 
                    "Robert Martin", 4);
    
    // Rebuild indices
    searchEngine->buildIndices();
    
    cout << "Sample data initialized: " << getTotalBooks() << " books added." << endl;
}

// ============ VALIDATION ============

bool LibraryManager::isISBNValid(const string& isbn) {
    // Basic ISBN validation (ISBN-10 or ISBN-13 format)
    if (isbn.length() < 10) return false;
    
    // Check for ISBN-13 format: 978-X-XXX-XXXXX-X
    if (isbn.find("978") == 0 || isbn.find("979") == 0) {
        return true;  // Simplified validation
    }
    
    return isbn.length() >= 10;  // Basic length check
}

bool LibraryManager::isBookAvailable(const string& isbn) {
    Book* book = bookTree->search(isbn);
    return (book != nullptr && book->isAvailable());
}

File: backend/libraryManager.js
// backend/libraryManager.js
const { loadBooks, saveBooks, loadUsers, saveUsers, loadTransactions, saveTransactions, generateTransactionID } = require('./dataManager');

const MAX_BORROW_LIMIT = 5;

// ============ BOOK OPERATIONS ============

function getAllBooks() {
    return loadBooks();
}

function getAvailableBooks() {
    const books = loadBooks();
    return books.filter(book => book.availableCopies > 0);
}

function getBookByISBN(isbn) {
    const books = loadBooks();
    return books.find(book => book.isbn === isbn);
}

function addBook(isbn, title, author, quantity) {
    const books = loadBooks();

    // Check if ISBN already exists
    if (books.find(book => book.isbn === isbn)) {
        return { success: false, message: 'Book with this ISBN already exists' };
    }

    const newBook = {
        isbn,
        title,
        author,
        quantity: parseInt(quantity),
        availableCopies: parseInt(quantity)
    };

    books.push(newBook);
    saveBooks(books);

    return { success: true, message: 'Book added successfully', book: newBook };
}

function removeBook(isbn) {
    const books = loadBooks();
    const index = books.findIndex(book => book.isbn === isbn);

    if (index === -1) {
        return { success: false, message: 'Book not found' };
    }

    // Check if book is borrowed
    if (books[index].availableCopies < books[index].quantity) {
        return { success: false, message: 'Cannot remove book that is currently borrowed' };
    }

    books.splice(index, 1);
    saveBooks(books);

    return { success: true, message: 'Book removed successfully' };
}

function updateBookDetails(isbn, newTitle, newAuthor) {
    const books = loadBooks();
    const book = books.find(b => b.isbn === isbn);

    if (!book) {
        return { success: false, message: 'Book not found' };
    }

    book.title = newTitle;
    book.author = newAuthor;
    saveBooks(books);

    return { success: true, message: 'Book updated successfully', book };
}

function updateBookQuantity(isbn, newQuantity) {
    const books = loadBooks();
    const book = books.find(b => b.isbn === isbn);

    if (!book) {
        return { success: false, message: 'Book not found' };
    }

    const borrowed = book.quantity - book.availableCopies;

    if (newQuantity < borrowed) {
        return { success: false, message: `Cannot set quantity below borrowed count (${borrowed})` };
    }

    book.availableCopies = newQuantity - borrowed;
    book.quantity = newQuantity;
    saveBooks(books);

    return { success: true, message: 'Quantity updated successfully', book };
}

// ============ SEARCH OPERATIONS ============

function searchBooksByTitle(title) {
    const books = loadBooks();
    const searchTerm = title.toLowerCase();
    return books.filter(book => book.title.toLowerCase().includes(searchTerm));
}

function searchBooksByAuthor(author) {
    const books = loadBooks();
    const searchTerm = author.toLowerCase();
    return books.filter(book => book.author.toLowerCase().includes(searchTerm));
}

function searchBooksByKeyword(keyword) {
    const books = loadBooks();
    const searchTerm = keyword.toLowerCase();
    return books.filter(book =>
        book.title.toLowerCase().includes(searchTerm) ||
        book.author.toLowerCase().includes(searchTerm) ||
        book.isbn.toLowerCase().includes(searchTerm)
    );
}

// ============ USER OPERATIONS ============

function getAllUsers() {
    return loadUsers();
}

function getUserByID(userID) {
    const users = loadUsers();
    return users.find(user => user.userID === userID);
}

function removeUser(userID) {
    const users = loadUsers();
    const index = users.findIndex(user => user.userID === userID);

    if (index === -1) {
        return { success: false, message: 'User not found' };
    }

    // Check if user has borrowed books
    if (users[index].borrowedISBNs.length > 0) {
        return { success: false, message: 'Cannot remove user with borrowed books' };
    }

    users.splice(index, 1);
    saveUsers(users);

    return { success: true, message: 'User removed successfully' };
}

function deactivateUser(userID) {
    const users = loadUsers();
    const user = users.find(u => u.userID === userID);

    if (!user) {
        return { success: false, message: 'User not found' };
    }

    user.isActive = false;
    saveUsers(users);

    return { success: true, message: 'User deactivated successfully' };
}

function activateUser(userID) {
    const users = loadUsers();
    const user = users.find(u => u.userID === userID);

    if (!user) {
        return { success: false, message: 'User not found' };
    }

    user.isActive = true;
    saveUsers(users);

    return { success: true, message: 'User activated successfully' };
}

function getActiveUsersCount() {
    const users = loadUsers();
    return users.filter(user => user.isActive).length;
}

// ============ BORROW/RETURN OPERATIONS ============

function borrowBook(isbn, userID) {
    const books = loadBooks();
    const users = loadUsers();

    const book = books.find(b => b.isbn === isbn);
    const user = users.find(u => u.userID === userID);

    if (!book) {
        return { success: false, message: 'Book not found' };
    }

    if (!user) {
        return { success: false, message: 'User not found' };
    }

    if (!user.isActive) {
        return { success: false, message: 'User account is inactive' };
    }

    if (book.availableCopies <= 0) {
        return { success: false, message: 'Book is not available' };
    }

    if (user.borrowedISBNs.length >= MAX_BORROW_LIMIT) {
        return { success: false, message: `Maximum borrow limit (${MAX_BORROW_LIMIT}) reached` };
    }

    if (user.borrowedISBNs.includes(isbn)) {
        return { success: false, message: 'You have already borrowed this book' };
    }

    // Update book availability
    book.availableCopies--;
    saveBooks(books);

    // Update user borrowed books
    user.borrowedISBNs.push(isbn);
    saveUsers(users);

    // Create transaction
    const transaction = {
        transactionID: generateTransactionID(),
        type: 'BORROW',
        userID: user.userID,
        userName: user.fullName,
        isbn: book.isbn,
        bookTitle: book.title,
        timestamp: new Date().toISOString()
    };

    const transactions = loadTransactions();
    transactions.push(transaction);
    saveTransactions(transactions);

    return { success: true, message: 'Book borrowed successfully', transaction };
}

function returnBook(isbn, userID) {
    const books = loadBooks();
    const users = loadUsers();

    const book = books.find(b => b.isbn === isbn);
    const user = users.find(u => u.userID === userID);

    if (!book) {
        return { success: false, message: 'Book not found' };
    }

    if (!user) {
        return { success: false, message: 'User not found' };
    }

    if (!user.borrowedISBNs.includes(isbn)) {
        return { success: false, message: 'You have not borrowed this book' };
    }

    // Update book availability
    book.availableCopies++;
    saveBooks(books);

    // Update user borrowed books
    user.borrowedISBNs = user.borrowedISBNs.filter(i => i !== isbn);
    saveUsers(users);

    // Create transaction
    const transaction = {
        transactionID: generateTransactionID(),
        type: 'RETURN',
        userID: user.userID,
        userName: user.fullName,
        isbn: book.isbn,
        bookTitle: book.title,
        timestamp: new Date().toISOString()
    };

    const transactions = loadTransactions();
    transactions.push(transaction);
    saveTransactions(transactions);

    return { success: true, message: 'Book returned successfully', transaction };
}

function getUserBorrowedBooks(userID) {
    const users = loadUsers();
    const books = loadBooks();

    const user = users.find(u => u.userID === userID);

    if (!user) {
        return [];
    }

    return user.borrowedISBNs.map(isbn => books.find(b => b.isbn === isbn)).filter(b => b);
}

// ============ TRANSACTION OPERATIONS ============

function getAllTransactions() {
    return loadTransactions();
}

function getUserTransactions(userID) {
    const transactions = loadTransactions();
    return transactions.filter(trans => trans.userID === userID);
}

function getBookTransactions(isbn) {
    const transactions = loadTransactions();
    return transactions.filter(trans => trans.isbn === isbn);
}

function getRecentTransactions(count) {
    const transactions = loadTransactions();
    return transactions.slice(-count).reverse();
}

// ============ STATISTICS ============

function getStatistics() {
    const books = loadBooks();
    const users = loadUsers();
    const transactions = loadTransactions();

    return {
        totalBooks: books.length,
        totalCopies: books.reduce((sum, book) => sum + book.quantity, 0),
        availableCopies: books.reduce((sum, book) => sum + book.availableCopies, 0),
        borrowedCopies: books.reduce((sum, book) => sum + (book.quantity - book.availableCopies), 0),
        totalUsers: users.length,
        activeUsers: users.filter(u => u.isActive).length,
        totalTransactions: transactions.length
    };
}

module.exports = {
    getAllBooks,
    getAvailableBooks,
    getBookByISBN,
    addBook,
    removeBook,
    updateBookDetails,
    updateBookQuantity,
    searchBooksByTitle,
    searchBooksByAuthor,
    searchBooksByKeyword,
    getAllUsers,
    getUserByID,
    removeUser,
    deactivateUser,
    activateUser,
    getActiveUsersCount,
    borrowBook,
    returnBook,
    getUserBorrowedBooks,
    getAllTransactions,
    getUserTransactions,
    getBookTransactions,
    getRecentTransactions,
    getStatistics
};
